### PR TITLE
Implement crc8 and hardware watchdog timer and radio wdt

### PR DESCRIPTION
**Pull request by yxyyeah**
This change is [Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6)

### CODE REVIEW

**Review by Rio6 — APPROVED** [review]
LGTM

**Review by itslinotlie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6)* status: 0 of 15 files reviewed, 1 unresolved discussion (waiting on @BluCodeGH and @yxyyeah)

___
*[`ltt_test/nbproject/Makefile-local-default.mk` line 14 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScRkEn2pHsD12nuQwV:-NScRkEp2tmhPoE8Tung:bh3uqhy) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/nbproject/Makefile-local-default.mk#L14)):*
> ```Makefile
> # least once so the file gets created and the project can be built. Finally, you can also
> # avoid using this file at all if you are only building from the command line with make.
> # You can invoke make with the values of the macros:
> ```

I noticed that cansw_telemetry has the same issue payload had (where the .gitignore isn't really ignoring the files we want it to ignore). The fix is to append ```**/``` in front of the nbproject files ie:

```
**/nbproject/private/
**/nbproject/Package-*.bash
**/nbproject/Makefile-*

```

and also removing the files from git (```git rm —cached <folder>/```) will do the trick

**Review by BluCodeGH — CHANGES_REQUESTED** [review]
Is the plan to have a separate PR for the radio watchdog?

> Reviewed 1 of 11 files at r1, 14 of 14 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-:-NScYFi_2_Q93w9Keoka:b-64gdab)* status: all files reviewed, 10 unresolved discussions (waiting on @yxyyeah)

___
*[`ltt_test/canlib` line 1 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NSabABE8BzoxtrvC0-B:-NSabABE8BzoxtrvC0-C:b-2ykrhz) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/canlib#L1)):*
> ```
> Subproject commit 3e4452b08dad6aecf56cc9dfa40cfea0c0b14a90
> ```

Make sure this gets updated to the latest master before merging
___
*[`ltt_test/main.c` line 137 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScYBr93rGtEA7rYNQc:-NScYBr93rGtEA7rYNQd:b-o8gb9l) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/main.c#L137)):*
> ```C
>     int dest_id = -1;
> 
>     switch (msg_type) {
> ```

Missing actuator message to let us manually toggle radio power?
___
*[`ltt_test/radio.c` line 48 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScX7nL-Xtk3yLMFL9x:-NScX7nL-Xtk3yLMFL9y:bbltgi2) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/radio.c#L48)):*
> ```C
>     static uint8_t parse_i = 0;
>     static can_msg_t msg;
>     static uint8_t crc_read1 = 0; // read the first hex of the parity byte
> ```

I'd just declare a single `crc_i` that acts like an index just like `parse_i`.
___
*[`ltt_test/radio.c` line 50 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScXDK36X-jmt987bd0:-NScXDK36X-jmt987bd1:b-l36bq5) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/radio.c#L50)):*
> ```C
>     static uint8_t crc_read1 = 0; // read the first hex of the parity byte
>     static uint8_t crc_read2 = 0; // read the second hex of the parity byte
>     static uint8_t exp_crc;
> ```

You don't need to declare every variable up here (this isn't C80), so I'd just define this in the little local area where its used.
___
*[`ltt_test/radio.c` line 67 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScXQJd3jppacuS5tAZ:-NScXQJd3jppacuS5tA_:b2f0y8g) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/radio.c#L67)):*
> ```C
>             return;
>         }
>         msg.sid |= (uint16_t) d << ((3 - parse_i) * 4);
> ```

Please add more parenthesis to clarify what's being cast here.
___
*[`ltt_test/radio.c` line 78 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScXZ6_5_Tag_pI8aPI:-NScXZ6_5_Tag_pI8aPJ:b-s710e9) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/radio.c#L78)):*
> ```C
>         }
>         if (c == ';') { // end of message
>             crc_read1 = 1;
> ```

missing `return`.
___
*[`ltt_test/radio.c` line 98 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScXb2D7Cycol_T7P3j:-NScXb2D7Cycol_T7P3k:byvbbgn) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/radio.c#L98)):*
> ```C
>         }
>         if (crc_read2) {
>             exp_crc = ((exp_crc << 4) & 0xf0) | (d & 0xf); // is it?
> ```

Yes, it is. However, your bitwise ands are unnecessary and make this harder to understand, so I'd remove them.
___
*[`ltt_test/radio.c` line 103 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScXk401KTX8noJAj-E:-NScXk401KTX8noJAj-F:b85rcnw) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/radio.c#L103)):*
> ```C
>             sid_arr[0] = (msg.sid) >> 8;
>             sid_arr[1] = (msg.sid) & 0xff;
>             uint8_t crc = crc8_checksum(sid_arr, sizeof (sid_arr), 0x00);
> ```

Probably easier to manually update the CRC with the two SID bytes rather than putting it in this array.
___
*[`ltt_test/radio.c` line 159 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScXtyS8-a2qMEu6yVI:-NScXtyS8-a2qMEu6yVJ:b-jb307v) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/radio.c#L159)):*
> ```C
>     sid_arr[0] = (msg->sid) >> 8;
>     sid_arr[1] = (msg->sid) & 0xff;
>     uint8_t crc = crc8_checksum(sid_arr, sizeof (sid_arr), 0x00);
> ```

Same comment as the earlier `sid_arr`.

**Review by itslinotlie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6)* status: all files reviewed, 10 unresolved discussions (waiting on @yxyyeah)

___
*[`ltt_test/nbproject/Makefile-local-default.mk` line 14 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScRkEn2pHsD12nuQwV:-NSfbMnW3p8XpJb2KxA_:b-55gznm) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/nbproject/Makefile-local-default.mk#L14)):*
<details><summary><i>Previously, itslinotlie (Michael Li) wrote…</i></summary><blockquote>

I noticed that cansw_telemetry has the same issue payload had (where the .gitignore isn't really ignoring the files we want it to ignore). The fix is to append ```**/``` in front of the nbproject files ie:

```
**/nbproject/private/
**/nbproject/Package-*.bash
**/nbproject/Makefile-*

```

and also removing the files from git (```git rm —cached <folder>/```) will do the trick


</blockquote></details>

I would like to retract my statement earlier. I think the cleaner solution would be to include .gitignore in the subdirectories (```ltt_test``` and ```ltr_test.X```) and the gitignore contents would be the same as the outer most .gitignore (still remove the files from git though)

**Review by BluCodeGH — COMMENTED** [review]
I realized we should also probably clean up this repository, since there is only LTT nowadays I think we can get rid of `ltr_test.X` and move everything out of `ltt_test` and into the repository root. Probably add a `tools/` folder and move `checksum.py` and `logging/main.py` into that.

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-:-NSg2XmC8Zb8N3qj1pDL:bfn4apj)* status: all files reviewed, 10 unresolved discussions (waiting on @itslinotlie and @yxyyeah)

___
*[`ltt_test/nbproject/Makefile-local-default.mk` line 14 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScRkEn2pHsD12nuQwV:-NSg2Q8i1g7GGnt7p7wG:b-89wjgf) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/nbproject/Makefile-local-default.mk#L14)):*
<details><summary><i>Previously, itslinotlie (Michael Li) wrote…</i></summary><blockquote>

I would like to retract my statement earlier. I think the cleaner solution would be to include .gitignore in the subdirectories (```ltt_test``` and ```ltr_test.X```) and the gitignore contents would be the same as the outer most .gitignore (still remove the files from git though)
</blockquote></details>

Actually I think you were right originally, in general we always want to ignore the nbproject files so we should add the `**/`.

**Review by yxyyeah — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6)* status: all files reviewed, 10 unresolved discussions (waiting on @BluCodeGH and @itslinotlie)

___
*[`ltt_test/main.c` line 137 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScYBr93rGtEA7rYNQc:-NSx5n_G-WibxXtcjyvq:b-s3vxz1) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/main.c#L137)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

Missing actuator message to let us manually toggle radio power?
</blockquote></details>

I think this is for bus power for the old code?

**Review by yxyyeah — COMMENTED** [review]
I made another branch for radio watchdog. So maybe later I can first merge that into this one

> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-:-NSxBnPl8v9h8AhzCUBn:b63lslp)* status: 14 of 15 files reviewed, 10 unresolved discussions (waiting on @BluCodeGH and @itslinotlie)

**Review by BluCodeGH — CHANGES_REQUESTED** [review]
Sounds good.

> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-:-NTB4yjqDLjEizABrw0_:brxp9sq)* status: all files reviewed, 3 unresolved discussions (waiting on @itslinotlie and @yxyyeah)

___
*[`ltt_test/main.c` line 137 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScYBr93rGtEA7rYNQc:-NTB4VI30ccLLC_3yvdI:bxxvcp9) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/main.c#L137)):*
<details><summary><i>Previously, yxyyeah (yxy) wrote…</i></summary><blockquote>

I think this is for bus power for the old code?
</blockquote></details>

I mean that we should be able to manually toggle power to the RFD900x so that if its having issues we can reboot it via hardwired CAN support. I guess that should happen in the other watchdog PR.
___
*[`ltt_test/radio.c` line 67 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScXQJd3jppacuS5tAZ:-NTB3XNW1_BGBa4kw4__:b-r2a2bk) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/radio.c#L67)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

Please add more parenthesis to clarify what's being cast here.
</blockquote></details>

I'm not sure if this works as written, since `d` is a `uint8_t` if we shift it by 8 bits (first hex digit) before casting to `uint16_t` won't that cause issues? I'd cast and then do the shift.

**Review by yxyyeah — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6)* status: 13 of 15 files reviewed, 3 unresolved discussions (waiting on @BluCodeGH and @itslinotlie)

___
*[`ltt_test/radio.c` line 67 at r3](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-NScXQJd3jppacuS5tAZ:-NTv5Pf06-TSPTGbOo3B:bb6un1j) ([raw file](https://github.com/waterloo-rocketry/cansw_telemetry/blob/49f6a2e8b77dd2f2da9465b94d319ae001324f05/ltt_test/radio.c#L67)):*
<details><summary><i>Previously, BluCodeGH (BluCode) wrote…</i></summary><blockquote>

I'm not sure if this works as written, since `d` is a `uint8_t` if we shift it by 8 bits (first hex digit) before casting to `uint16_t` won't that cause issues? I'd cast and then do the shift.
</blockquote></details>

Done

**Review by BluCodeGH — COMMENTED** [review]
.gitignore still needs to be updated (See Michael's early comment), otherwise :lgtm: 

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/waterloo-rocketry/cansw_telemetry/6#-:-NUEfjss9oiLeK6ULGmN:br1t954)* status: all files reviewed, 2 unresolved discussions (waiting on @itslinotlie and @yxyyeah)

File: ltt_test/radio.h [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 #include "canlib.h"
 
-void radio_handle_input_character(char c);
+bool radio_handle_input_character(char c);
```
Comment (Rio6) [inline ltt_test/radio.h]: I think this got missed when reverting the function signature

File: ltt_test/main.c [hunk 2]
```diff
@@ -62,144 +59,116 @@ int main(void) {
...
+                    radio_power = false;
+                }
+                else if (get_req_actuator_state(msg) == ACTUATOR_ON && !radio_power){
+                    set_radio_power = true;
+                    radio_power = true;
+                }
             }
             break;
-
+
```
Comment (Rio6) [inline ltt_test/main.c]: missed trailing space

File: ltt_test/main.c [hunk 3]
```diff
@@ -62,144 +59,116 @@ int main(void) {
...
-                req_bus_powered = get_req_actuator_state(msg) == ACTUATOR_CLOSED;
+            if (get_actuator_id(msg) == ACTUATOR_RADIO) {
+                if (get_req_actuator_state(msg) == ACTUATOR_OFF && radio_power){
+                    set_radio_power = true;
+                    radio_power = false;
+                }
+                else if (get_req_actuator_state(msg) == ACTUATOR_ON && !radio_power){
+                    set_radio_power = true;
+                    radio_power = true;
+                }
```
Comment (Rio6) [inline ltt_test/main.c]: might be a good idea to remove check on radio_power so the command always set the power to desired one regardless of the previous state, for cases where there is a bug or a hyperintelligent cosmic proton messes with the state.

File: ltt_test/main.c [hunk 4]
```diff
@@ -62,144 +59,116 @@ int main(void) {
...
         //send any queued CAN messages
         txb_heartbeat();
+        
+        // clear watchdog timer
+        CLRWDT();
+        
+        // reset radio if no message received for over 1 minute
+        if (millis() - last_message_millis > RESET_RADIO_ms){
+            // reset radio
+            SET_RADIO_POWER(0);
```
Comment (Rio6) [inline ltt_test/main.c]: we can possibly send a error state message to CAN when this happens

**Comment by Rio6** [comment]
Changed telemetry into binary format.

Depends on https://github.com/waterloo-rocketry/parsley/pull/4

Depends on https://github.com/waterloo-rocketry/omnibus/pull/114

**Review by BluCodeGH — APPROVED** [review]
Flow (twice), lets get this merged.

**Pull request merged**